Remove 2fa requirement for mobile users

proteusvacuum opened the pull request:
We currently have a security option that says `All web users on this project will be required to enable two factor authentication`. However, turning on this toggle required _all_ users (including mobile workers) to have 2FA enabled. We don't have a mechanism for mobile workers to enter a 2fa code from the phone, so this essentially prevents them from being able to do anything on HQ. 

I tested turning on this option and
- Loading a page as a web user: this gave me the "you require 2fa" page
- Loading the same page logged in as a mobile worker: this let me load the page.

I think this is a little funny, although I'm assuming the intention was that web users can potentially have more power (despite mobile workers having the ability to do everything web users do if they have the right role...).

I think this might need a little more thought, as this just seems wrong to me, but I think was the original intention of this option.

https://manage.dimagi.com/default.asp?269293#1451924

proteusvacuum committed:
Remove 2fa requirement for mobile users

emord commented:
I haven't done a deep dive into this, but I imagine this could have been broken by the recent auth decorator refactor.

> Loading a page as a web user: this gave me the "you require 2fa" page
> Loading the same page logged in as a mobile worker: this let me load the page.
 
I think we may want to let you always restore/submit forms/download apps without requiring two factor, but everything else should require two factor. Some mobile workers have access to reports on the web and that should likely also require two factor (ICDS does this for dashboard and I think some other projects do similar things).

krsdimagi commented:
@proteusvacuum @emord when we first thought about implementing this, it was in fact just for web users.  I think if we want to have a mobile two factor option we should do some more thinking about how we want to roll that out.

Additionally, we are going to be rolling out a blog post shortly that highlights security features such as two factor, and in its current state it only discusses web users - it's not a bad idea to eventually roll out some sort of two factor on mobile, but I think we might want to be careful with timing.

proteusvacuum commented:
@krsdimagi have we thought about having 2FA based on role instead of user type? I think the distinction between web users and mobile users is actually moot since we can now attach roles to mobile users (the only meaningful distinction is that web users can't log in on the mobile). This means that mobile users with highly privileged roles could not be required to have 2FA enabled, which seems like a big security hole to me.

krsdimagi commented:
@proteusvacuum might we be able to build out 2FA configuration for mobile users instead?  I think having it based on role might get a bit complicated, and I think it would be nice to tell users that if 2FA is turned on for all users, that means any user with HQ access.

proteusvacuum commented:
@krsdimagi I think that's what @emord was suggesting above: have 2FA auth enabled for every URL on HQ, but not for anything that the mobile uses. That sounds closer to "correct", since I'm assuming we've already ruled out some way of entering the 2FA token in the mobile app?

krsdimagi commented:
@proteusvacuum Cool, that makes sense to me.  At some point I think it will be good to think about this for the mobile app, but because 2FA is a feature that has been around for a while I think it is probably good to get the HQ side taken care of right now.

esoergel commented:
Not much to add here, but I wanted to flag that the `@mobile_auth` decorator is expressly intended for views meant for mobile devices, although there are plenty of mobile endpoints that use some other means of authentication (`receiverwrapper.views` is pretty custom, for instance).  @proteusvacuum what you are proposing could potentially be rolled in to that decorator, since it circumvents the normal auth channels anyways.  I haven't looked at converting the other mobile views to that decorator though.

(I do think the app manager views are largely unauthenticated, and intentionally so, as you have to download the app before you can authenticate)

czue reviewed (changes requested):
I also don't have a lot to add to the conversation here, I agree with the conclusion @emord and @proteusvacuum have come to. 

Sounds like there will be some changes before this gets merged, yeah? Requesting changes assuming that's the case.

proteusvacuum committed:
Prefactor -> remove one level of indentation

proteusvacuum committed:
Refactor -> Group auth functions together

proteusvacuum committed:
Pass allow_two_factor kwarg

proteusvacuum committed:
Revert "Pass allow_two_factor kwarg"

This reverts commit 3636d8a8c821b4cea2d26dd82d23892eb39aae84.

proteusvacuum committed:
Use function attribute to determine 2FA availability

proteusvacuum committed:
Update help text for two factor auth setting

proteusvacuum committed:
Use decorator to mark two factor exemption

Taking the same pattern as @csrf_exempt from django

proteusvacuum committed:
Decorate two_factor exempt views

proteusvacuum committed:
Update docs

proteusvacuum commented:
@emord @esoergel @czue 
I've updated this PR now with a more comprehensive strategy. I went in a few directions, as you can see from the commit history, but this now adds a `two_factor_exempt` decorator that is automatically applied by the `mobile_auth` decorators. I've also added this to the `receiverwrapper.views.secure_post`, as I believe that endpoint should also be exempt. 

🐡 / 🐡 from https://github.com/dimagi/commcare-hq/pull/19289/commits/d3db8ca7ec6cdeef0e28ff9442869a4551690e99

This is the list of all the endpoints that have that decorator applied:
```
secure_post
restore
search
claim
heartbeat
get_next_id
fetch_key_records
migration_restore

## eNikshay specific, this domain doesn't have that option turned on, but added just to be safe
update_patient_adherence
update_patient_details
update_adherence_confidence
update_default_confidence
payment_confirmation
```

Since this is another significant auth change (and 2FA update), I'm going to stick it on staging and ask for a quick QA pass.

proteusvacuum committed:
Fix function return for api auth

https://manage.dimagi.com/default.asp?269999

proteusvacuum commented:
https://github.com/dimagi/commcare-hq/pull/19289/commits/0487352866dd65e681ed4584c5eb03cb826d3258 fixes a different bug with 2FA I discovered while testing this out: https://manage.dimagi.com/default.asp?269999

proteusvacuum commented:
BTW, I'm still a bit confused about how this works for API endpoints - we currently require any API requests to include a `X-CommcareHQ-OTP` header with the current "token", which seems like a funny way to access an API. 
Not sure what people think about this, or what the original intention was?

proteusvacuum commented on corehq/apps/domain/decorators.py:
```diff
@@ -80,51 +80,52 @@ def login_and_domain_required(view_func):
     def _inner(req, domain, *args, **kwargs):
         user = req.user
         domain_name, domain = load_domain(req, domain)
-        if domain:
```
view this commit with https://github.com/dimagi/commcare-hq/pull/19289/commits/d3db8ca7ec6cdeef0e28ff9442869a4551690e99?w=1

czue reviewed (approved):
:clap: this looks great to me. i like the approach you landed on.

czue commented:
and sorry, i've got no context on the existing system, but agree that application-specific tokens would be nice. does this have anything to do with the `api_key` thing that we also have or is that a totally different thing?

proteusvacuum commented:
@czue yeah, currently if you use "API key" auth, it bypasses the requirement for 2FA. I haven't been part of the greater 2FA discussions, so I'm not sure if this was already thought about, but is the theory behind having the API key as the "password" for api endpoints that you need to be able to log in to HQ to obtain the key first (i.e. need the 2 factors first)?

proteusvacuum commented:
also pinging @pr33thi as she's working on 2FA now.

esoergel commented on corehq/apps/domain/decorators.py:
```diff
@@ -287,8 +293,8 @@ def _inner(request, domain, *args, **kwargs):
-def _two_factor_required(domain, couch_user):
-    return domain.two_factor_auth and couch_user.is_web_user() and not couch_user.two_factor_disabled
+def _two_factor_required(allow_two_factor, domain, couch_user):
+    return allow_two_factor and domain.two_factor_auth and not couch_user.two_factor_disabled
```
nit: I was a little confused about what `allow_two_factor` means - I think it's "this view requires two factor if enabled on the domain", but that's hard to fit into a variable name.  Another option would be (`not`) `skip_two_factor`.  Take your pick.

esoergel commented on corehq/apps/domain/decorators.py:
```diff
@@ -231,6 +231,8 @@ def login_or_api_key_ex(allow_cc_users=False, allow_sessions=True):
 
 def _get_multi_auth_decorator(default, allow_token=False, allow_two_factor=True):
     def decorator(fn):
+        fn.allow_two_factor = allow_two_factor
```
I like this approach.  Django also does some magic such that as long as you decorate the `dispatch` method, it'll copy this attr over to the generated view function.

czue commented:
> is the theory behind having the API key as the "password" for api endpoints that you need to be able to log in to HQ to obtain the key first (i.e. need the 2 factors first)?

i don't know the answer to this, nor do I know if it was fully thought of (guessing maybe we just developed api auth and 2FA independently and never thought about the interplay between the two).

my quick opinions on this (not trying to be prescriptive about what we should do so much as just think out loud):

it seems like an accurate state of the world is: "enabling 2FA does not affect API access via tokens", and if we wanted domains to be able to control API access via 2FA we'd likely have to just disable it for any 2FA domains (or provide a setting to do so). that feels like the "responsible" thing to do assuming a goal of 2FA is data protection, since an API token can get you access to an entire domain's data, though i also recognize that could add quite a bit more work/complexity to what is already a rather complex set of rules governing auth.

esoergel commented on corehq/apps/domain/decorators.py:
```diff
@@ -249,13 +247,29 @@ def _inner(request, *args, **kwargs):
+    def wrapped_view(*args, **kwargs):
+        return view_func(*args, **kwargs)
+    wrapped_view.two_factor_exempt = True
+    return wraps(view_func, assigned=available_attrs(view_func))(wrapped_view)
```
Ah, is this the distinction?  (Checking my understanding)

In the common usage, they should behave identically:
```python
@two_factor_exempt
def view_func(request):
    return "hi"
```
But the simpler implementation mutates the function, so if you needed access to the wrapped and unwrapped versions:
```python
def view_func(request):
    return "hi"

two_factor_exempt_view_func = two_factor_exempt(view_func)
```
Then both `view_func` and `two_factor_exempt_view_func` would both (surprisingly) have the attr set?  That's quite a gotcha.

esoergel commented on corehq/apps/domain/decorators.py:
```diff
@@ -249,13 +247,29 @@ def _inner(request, *args, **kwargs):
 
+def two_factor_exempt(view_func):
+    """
+    Marks a view function as being exempt from two factor authentication.
```
Nice, I like this framing-in-the-negative better.

esoergel reviewed (approved):
Cool, I like where you landed with this.  I assume you've communicated with product about this, since I think they think two-factor is web user only.  @pr33thi may also be interested, since she's doing some two-factor stuff now too.

snopoke commented on corehq/apps/domain/decorators.py:
```diff
@@ -80,51 +80,52 @@ def login_and_domain_required(view_func):
-                    else:
-                        return view_func(req, domain_name, *args, **kwargs)
+        if not domain:
+            msg = _(('The domain "{domain}" was not found.').format(domain=domain_name))
```
nit: in a user facing message I think the term for a 'domain' is a 'project'. Same for other places.

proteusvacuum replied:
I actually didn't add this, just removed one level of indentation here. I'm going to leave this as is for now.

snopoke reviewed (commented):
not much to add to what others have said except that our auth is complicated!

mkangia committed:
Merge branch 'master' into fr/two-factor-mobile-users

mkangia commented:
@proteusvacuum was facing a conflict with staging deploy for this branch. Resolved in last commit.

calellowitz committed:
Merge branch 'master' into fr/two-factor-mobile-users

calellowitz commented:
@proteusvacuum resolved a merge conflict to deploy staging as well

proteusvacuum committed:
Properly wrap secure_post sub-methods

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/two-factor-mobile-users

proteusvacuum commented:
Passed QA :+1:

calellowitz commented:
Definitely late to the party on this one, but the original intent was to have the exceptions for the URLs, as emord said, not all mobile users. My guess is that it this was broken during one of the many changes we have made to the auth decorators since then (or never worked but that seems less likely, since I imagine syncing was part of the original QA). Either way, thanks for fixing this @proteusvacuum
